modules/fx/arrow-fx-kotlinx-coroutines/src/main/kotlin/arrow/integrations/kotlinx/extensions.kt

```diff
@@ -0,0 +1,37 @@
+package arrow.integrations.kotlinx
+
+import arrow.core.Either
+import arrow.fx.IOOf
+import arrow.fx.OnCancel
+import arrow.fx.fix
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.newCoroutineContext
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlin.coroutines.EmptyCoroutineContext
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+
+suspend fun <A> IOOf<A>.suspendCancellable() = suspendCancellableCoroutine<A> { cont ->
+  val disposable = fix().unsafeRunAsyncCancellable(OnCancel.ThrowCancellationException) { result ->
+    result.fold(cont::resumeWithException) { cont.resume(it) }
+  }
+
+  cont.invokeOnCancellation { disposable() }
+}
+
+fun <A> IOOf<A>.unsafeRunScoped(
+  scope: CoroutineScope,
+  f: (Either<Throwable, A>) -> Unit
+): Unit {
+  val newContext = scope.newCoroutineContext(EmptyCoroutineContext)
+  val job = newContext[Job]
+
+  val disposable = fix().unsafeRunAsyncCancellable(cb = f)
+
+  job?.invokeOnCompletion { e ->
+    if (e is CancellationException) disposable()
+    else Unit
+  }
```

Comment:
Suggestion, add the opposite extfun:

```
fun CoroutineScope.unsafeRunIO(io: IOOf<A>, f: (Either<Throwable, A>) -> Unit)
```

that way it can be used directly from a CoroutineScope without the explicit argument.

Reply:
You don't really get rid of the argument, but get the syntax the other way around:

```
IO.just(i).unsafeRunScoped(scope) { ... }

scope.unsafeRunIO(IO.just(i)) { ... }
```

But fine with me anyway 👍

Reply:
Most `CoroutineScope`s I've seen in the wild are implemented by the class, so

```
class MyFragment: ScopedFragment {
  override fun onCreate() {
    runScopedIO(IO.just(1))
  }
}
```

Reply:
Ahh I see, awesome 👍